internal/build/lifecycle_execution.go

```diff
@@ -720,6 +740,67 @@ func (l *LifecycleExecution) ExtendRun(ctx context.Context, kanikoCache Cache, p
 	return extend.Run(ctx)
 }
 
+const (
+	argBuildID = "build_id"
+	argUserID  = "user_id"
+	argGroupID = "group_id"
+)
+
+/*
+	This implementation of ExtendBuildByDaemon is based on the RFC #0105 which uses docker daemon to extend the build Image instead of kaniko.
+	* Parsing the `group.toml` from the temp directory of buildpack and set the extensions.
+	* Reading the dockerfiles that were generated during the `generate` phase and also parsing the Arguments given by the user from
+	`extend-config.toml`.
+	* Using ImageBuild method of docker API client to extend the Image and save it to the daemon.
+	* Invoking Build phase of lifecycle by creating a container from the extended Image and dropping the privileges.
+
+*/
+
+func (l *LifecycleExecution) ExtendBuildByDaemon(ctx context.Context) error {
+	builderImageName := l.opts.BuilderImage
+	extendedBuilderImageName := l.opts.BuilderImage + "-extended"
+	defaultFilterFunc := func(file string) bool { return true }
+	var extensions Extensions
+	extensions.SetExtensions(l.tmpDir, l.logger)
+	dockerfiles, err := extensions.DockerFiles(DockerfileKindBuild, l.tmpDir, l.logger)
+	if err != nil {
+		return fmt.Errorf("getting %s.Dockerfiles: %w", DockerfileKindBuild, err)
+	}
+	for _, dockerfile := range dockerfiles {
+		dockerfile.Args = append([]Arg{
+			{Name: argBuildID, Value: uuid.New().String()},
+			{Name: argUserID, Value: strconv.Itoa(l.opts.Builder.UID())},
+			{Name: argGroupID, Value: strconv.Itoa(l.opts.Builder.GID())},
+		}, dockerfile.Args...)
+		buildContext := archive.ReadDirAsTar(filepath.Dir(dockerfile.Info.Path), "/", 0, 0, -1, true, false, defaultFilterFunc)
+		buildArguments := map[string]*string{}
+		buildArguments["base_image"] = &builderImageName
+		for i := range dockerfile.Args {
+			arg := &dockerfile.Args[i]
+			buildArguments[arg.Name] = &arg.Value
+		}
```

Comment:
```suggestion
		for _, arg := range dockerfile.Args {
			buildArguments[arg.Name] = &arg.Value
		}
```
This might read better

Reply:
no, actually i did this bcs ideally we shouldn't export a loop variable's pointer so made a copy of the loop variable